mono/mini/mini-llvm.c

```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
+				const int mask_values [] = { 8, 24, 9, 25, 10, 26, 11, 27, 12, 28, 13, 29, 14, 30, 15, 31 };
+				LLVMValueRef shuffled = LLVMBuildShuffleVector (builder, 
+					convert (ctx, lhs, type_to_simd_type (MONO_TYPE_I1)), 
+					convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I1)), 
+					create_const_vector_i32 (mask_values, 16), "");
+				values [ins->dreg] = convert (ctx, shuffled, type_to_simd_type (ins->inst_c1));
+			} else {
+				g_assert_not_reached ();
+			}
+			break;
+		}
+
+		case OP_SSE_LOADU: {
+			LLVMValueRef dst_ptr = convert (ctx, lhs, LLVMPointerType (primitive_type_to_llvm_type (ins->inst_c1), 0));
+			LLVMValueRef dst_vec = LLVMBuildBitCast (builder, dst_ptr, LLVMPointerType (type_to_simd_type (ins->inst_c1), 0), "");
+			values [ins->dreg] = mono_llvm_build_aligned_load (builder, dst_vec, "", FALSE, ins->inst_c0); // inst_c0 is alignment
+			break;
+		}
+
+		case OP_SSE_STORE: {
+			LLVMValueRef dst_vec = convert (ctx, lhs, LLVMPointerType (LLVMTypeOf (rhs), 0));
+			mono_llvm_build_aligned_store (builder, rhs, dst_vec, FALSE, ins->inst_c0);
+			break;
+		}
+
+		case OP_SSE_STORES: {
+			LLVMValueRef first_elem = LLVMBuildExtractElement (builder, rhs, LLVMConstInt (LLVMInt32Type (), 0, FALSE), "");
+			LLVMValueRef dst = convert (ctx, lhs, LLVMPointerType (LLVMTypeOf (first_elem), 0));
+			mono_llvm_build_aligned_store (builder, first_elem, dst, FALSE, 1);
+			break;
+		}
+
+		case OP_SSE_SHUFFLE: {
+			LLVMValueRef shuffle_vec = create_const_vector_4_i32 (
+				((ins->inst_c0 >> 0) & 0x3) + 0, // take two elements from lhs
+				((ins->inst_c0 >> 2) & 0x3) + 0, 
+				((ins->inst_c0 >> 4) & 0x3) + 4, // and two from rhs
+				((ins->inst_c0 >> 6) & 0x3) + 4);
+			values [ins->dreg] = LLVMBuildShuffleVector (builder, lhs, rhs, shuffle_vec, "");
+			break;
+		}
+
+		case OP_SSE2_SHUFFLE: {
+			LLVMValueRef right_vec;
+			LLVMValueRef shuffle_vec;
+			if (ins->inst_c1 == MONO_TYPE_R8) {
+				right_vec = rhs;
+				shuffle_vec = create_const_vector_2_i32 (
+					((ins->inst_c0 >> 0) & 0x1) + 0,
+					((ins->inst_c0 >> 1) & 0x1) + 2);
+			} else {
+				right_vec = LLVMGetUndef (LLVMVectorType (LLVMInt32Type (), 4));
+				shuffle_vec = create_const_vector_4_i32 (
+					(ins->inst_c0 >> 0) & 0x3,
+					(ins->inst_c0 >> 2) & 0x3, 
+					(ins->inst_c0 >> 4) & 0x3,
+					(ins->inst_c0 >> 6) & 0x3);
+			}
+			values [ins->dreg] = LLVMBuildShuffleVector (builder, lhs, right_vec, shuffle_vec, "");
+			break;
+		}
+
+		case OP_SSE_OR: {
+			LLVMValueRef minus_one [2];
+			minus_one [0] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
+			minus_one [1] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
+			LLVMValueRef vec_lhs_i64 = convert (ctx, lhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_rhs_i64 = convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_and = LLVMBuildOr (builder, vec_lhs_i64, vec_rhs_i64, "");
+			values [ins->dreg] = LLVMBuildBitCast (builder, vec_and, type_to_simd_type (ins->inst_c1), "");
```

Comment:
Can't this be handled by OP_XBINOP ?

Reply:
partially, won't work for floats/doubles.